website/blog/2020-03-02-presto-lambda.md

```diff
@@ -0,0 +1,194 @@
+- However, we should use the same class representing captured lambda for every `BIND` call, otherwise we will generate too many classes.
+
+We use the same approach as Java handles lambda and capture [3, 4]:
+- A lambda is represented as an object whose type is an functional interface (a.k.a. Single Abstract Method class)
+- The `invokedynamic` instruction is used to perform lambda capture:
+    * During the first `invokedynamic` call, the class representing the captured lambda (which is a functional interface) is created and a method to perform the capture is generated in it. This step is also called linkage and will only be done once.
+    * Every `invokedynamic` call performs capture and returns an instance of the desired functional interface. 
+
+With this design, `BIND` function will always be fused together with the lambda generation step to generate captured lambda in a single step -- we cannot first generate an object representing the uncaptured lambda, and then perform a separate partial application step. Note this implementation also doesn't allow more general higher-order function that returns a function as result.
+
+## Lambda in Aggregation 
+
+
+While lambda was originally introduced to help query nested data with scalar functions, we also noted it can be used in aggregation functions to allow more flexible analytics. The initial support for lambda in aggregation was added in [Presto#12084](https://github.com/prestodb/presto/pull/12084), with a [reduce_agg](https://prestodb.io/docs/current/functions/aggregate.html#reduce_agg) function for demonstration purpose. `reduce_agg` *conceputally* allows the creation of User-Defined Aggregation Function (UDAF) by making the input and combine functions lambdas. The following example shows how to use `reduce_agg` to compute group-wise product (instead of sum):
+
+```sql
+SELECT id, 
+reduce_agg(value, 1, (a, b) -> a * b, (a, b) -> a * b) prod
+FROM (
+    VALUES
+        (1, 2),
+        (1, 3),
+        (1, 4),
+        (2, 20),
+        (2, 30),
+        (2, 40)
+) AS t(id, value)
+GROUP BY id;
+
+ id | prod
+----+-------
+  2 | 24000
+  1 |    24
+(2 rows)
+```
+
+Unfortunately, due to [JDK-8017163](https://bugs.openjdk.java.net/browse/JDK-8017163), aggregation state with `Slice` or `Block` as native container type is intentionally not supported yet, as it may result in excessive JVM memory usage of remembered set. This is because aggregation states requires update in unpredictable ordering, results in a huge amount of cross-region references when each state is an separate object. This issue is also reported in [Presto#9553](https://github.com/prestodb/presto/issues/9553). This makes this function not yet practically useful. Once [JDK-8017163](https://bugs.openjdk.java.net/browse/JDK-8017163) is fixed in later versions of the JVM, we are looking forward to enabling it with more general types to allow more flexible analytics in aggregations!
```

Comment:
```
Unfortunately, due to [JDK-8017163](https://bugs.openjdk.java.net/browse/JDK-8017163), aggregation state with `Slice` or `Block` as a native container type is intentionally not supported yet. It can result in excessive JVM remembered set memory usage. This is because aggregation state requires updates in unpredictable order, resulting in a huge amount of cross-region references when each state is a separate object. This issue is also reported in [Presto#9553](https://github.com/prestodb/presto/issues/9553). This makes this function not yet practically useful. Once [JDK-8017163](https://bugs.openjdk.java.net/browse/JDK-8017163) is fixed in later versions of the JVM, we are looking forward to enabling it with more general types to allow more flexible analytics in aggregations!
```